### PR TITLE
rgw: some cleanup and fixes for librgw_file unittest

### PR DESCRIPTION
**Pull request by guihecheng**
Try to do cleanup works when --do_create && --do_delete specified in order to make it possible to run multiple-times continuously.

### CODE REVIEW

**Comment by mattbenjamin** [comment]
retest this please

File: src/test/librgw_file_aw.cc [hunk 1]
```diff
@@ -273,6 +273,11 @@ TEST(LibRGW, DELETE_OBJECT) {
     int ret = rgw_unlink(fs, bucket_fh, object_name.c_str(),
 			 RGW_UNLINK_FLAG_NONE);
     ASSERT_EQ(ret, 0);
+    if (do_create) {
```
Comment (mattbenjamin) [inline src/test/librgw_file_aw.cc]: should it do this?  I'm guessing it should be in if (do_delete)?

File: src/test/librgw_file_nfsns.cc [hunk 2]
```diff
@@ -221,7 +221,9 @@ TEST(LibRGW, SETUP_HIER1)
     if (do_create) {
       /* create objects directly */
       std::vector<std::string> obj_names =
-	{"foo/bar/baz/quux",
```
Comment (mattbenjamin) [inline src/test/librgw_file_nfsns.cc]: I think this is incorrect;  one of the behaviors we need to test is the ability of librgw_file to expand path segments including "virtual" directories;  this change removes an intentional implied path sequence by creating all the segments

File: src/test/librgw_file_nfsns.cc [hunk 3]
```diff
@@ -707,19 +707,40 @@ TEST(LibRGW, READ_DIRS1)
 
 TEST(LibRGW, READF_SETUP1)
 {
```
Comment (mattbenjamin) [inline src/test/librgw_file_nfsns.cc]: ok

File: src/test/librgw_file_nfsns.cc [hunk 4]
```diff
@@ -471,6 +472,11 @@ TEST(LibRGW, SETATTR) {
       ASSERT_EQ(st.st_uid, magic_uid);
       ASSERT_EQ(st.st_gid, magic_gid);
 
+      if (do_delete) {
```
Comment (mattbenjamin) [inline src/test/librgw_file_nfsns.cc]: ok

File: src/test/librgw_file_nfsns.cc [hunk 5]
```diff
@@ -623,19 +623,15 @@ TEST(LibRGW, BAD_DELETES_DIRS1) {
       rc = rgw_unlink(fs, dirs1_b.parent_fh, dirs1_b.name.c_str(),
 		      RGW_UNLINK_FLAG_NONE);
       ASSERT_NE(rc, 0);
+      /* try to unlink a non-empty directory (non-bucket) */
```
Comment (mattbenjamin) [inline src/test/librgw_file_nfsns.cc]: ok

File: src/test/librgw_file_aw.cc (hunk 1 above)
Comment (guihecheng) [inline src/test/librgw_file_aw.cc]: @mattbenjamin actually, here is alreadly inside an if (do_delete) {} block, this check is to see whether the bucket is created by '--create' option automatically, if not we'd better keep it there.

**Comment by guihecheng** [comment]
@mattbenjamin ok, I think I got it, I agree that we shall not cleanup sparse hierarchies.

**Comment by guihecheng** [comment]
@mattbenjamin updated, now the sparse hierarchies are not removed by --delete.

File: src/test/librgw_file_nfsns.cc (hunk 2 above)
Comment (guihecheng) [inline src/test/librgw_file_nfsns.cc]: @mattbenjamin hmmm... then there is some other problems that lead to ENOENT, will check it.

**Comment by guihecheng** [comment]
@mattbenjamin updated, I think the ENOENT from stat_leaf() is correct, and to "bridge" the "virtual" directories we should set the RGW_LOOKUP_FLAG_CREATE for rgw_lookup.

**Comment by mattbenjamin** [comment]
@guihecheng responding to the above comment on non-leaf, non-materialized directories, no, I don't agree with this.  These directories really do exist, and need to be returned from lookup, even if we don't have a leaf object for them.  FLAG_CREATE is to indicate that we are creating or instantiating something that may not exist--these directories do exist;  the only distinction is they don't have normal attributes until attrs are stored on them.

**Comment by mattbenjamin** [comment]
@guihecheng what I was trying to say about sparse hierarchies was that --create should still create them, but, --delete -should- delete them, if specified, no?

**Comment by mattbenjamin** [comment]
@guihecheng otoh, I am open to suggestions on how to better handle implied directories--in particular, their ownership and permissions.  I don't want to be forced to do create such dirs, esp. synchronously, though.  I thought about chasing parents to find a non-empty permission set.  Or allowing a "default permission set" to be defined.  What do you think?

**Comment by guihecheng** [comment]
@mattbenjamin oh, sorry, I had miss understand the "remove creation" on your comment on sparse hierarchies, so I shall do cleanup with --delete specified, sure.
And for ENOENT return for virtual directories, I think you've pointed out the cause indeed -- the pemission set; the "default pemission set" sound just enough to work; chasing parents for permission set sounds sane (do it implies that we are going to implement sth like permission inheritance on fh creation ?)

**Comment by mattbenjamin** [comment]
@guihecheng I hadn't thought of propagating perms on handle create, but it makes sense and has precedent;  I was going to say we would chase perms at runtime, so we wouldn't have to deal with propagating new perms when they are applied (e.g., someone realizes that /nfs has bogus owner:group, fixes it--we need to visit everything in cache to see if it needs new perms?)  it seems like we need a default or fallback permission at the top of the hierarchy in any case, as it is the base case?

**Comment by guihecheng** [comment]
@mattbenjamin ok, IMO, fallback permission sounds a more fs-like behaviour ;)

**Comment by guihecheng** [comment]
@mattbenjamin maybe we could also define a default permission to prevent worst cases -- fallback way failes (never happens ? ).

**Comment by guihecheng** [comment]
@mattbenjamin I've added back cleanups for do_hier1 with --delete and I add flag RGW_LOOKUP_FLAG_RCB to rgw_lookup() for do_hier1, because it is a lookup after readdir, we should specify RGW_LOOKUP_FLAG_RCB just as nfs-ganesha FSAL_RGW does.

**Comment by guihecheng** [comment]
@mattbenjamin rebased, thanks, an extra commit to handle special case for librgw_file_aw --large.

**Comment by guihecheng** [comment]
@mattbenjamin any more interest on this ?

**Comment by mattbenjamin** [comment]
@guihecheng I do want this, need to set aside time to test and automate--we'll help with the latter

**Comment by guihecheng** [comment]
@mattbenjamin oh, nice to hear that!

**Comment by mattbenjamin** [comment]
unstale me, my man

**Pull request closed without merging**